src/controlplane/controlplane_test.go

```diff
@@ -0,0 +1,310 @@
+	const defaultNamespace = "kube-system"
+
+	return map[controlplane.ComponentName]config.AutodiscoverControlPlane{
+		controlplane.Etcd: {
+			Namespace: defaultNamespace,
+			MatchNode: true,
+			Selector:  "k8s-app=etcd-manager-main",
+			Endpoints: []config.Endpoint{
+				{URL: server.ControlPlaneEndpoint(string(controlplane.Etcd))},
+			},
+		},
+		controlplane.APIServer: {
+			Namespace: defaultNamespace,
+			MatchNode: true,
+			Selector:  "k8s-app=kube-apiserver",
+			Endpoints: []config.Endpoint{
+				{URL: server.ControlPlaneEndpoint(string(controlplane.APIServer))},
+			},
+		},
+		controlplane.Scheduler: {
+			Namespace: defaultNamespace,
+			MatchNode: true,
+			Selector:  "k8s-app=kube-scheduler",
+			Endpoints: []config.Endpoint{
+				{URL: server.ControlPlaneEndpoint(string(controlplane.Scheduler))},
+			},
+		},
+		controlplane.ControllerManager: {
+			Namespace: defaultNamespace,
+			MatchNode: true,
+			Selector:  "k8s-app=kube-controller-manager",
+			Endpoints: []config.Endpoint{
+				{URL: server.ControlPlaneEndpoint(string(controlplane.ControllerManager))},
+			},
+		},
+	}
+}
+
+func testConfig(
+	autodiscovery map[controlplane.ComponentName]config.AutodiscoverControlPlane,
+	nodeName string,
+) config.Config {
+	return config.Config{
+		NodeName: nodeName,
+		ControlPlane: config.ControlPlane{
+			Enabled: true,
+			ETCD: config.ControlPlaneComponent{
+				Enabled: true,
+				Autodiscover: []config.AutodiscoverControlPlane{
+					autodiscovery[controlplane.Etcd],
+				},
+			},
+			APIServer: config.ControlPlaneComponent{
+				Enabled: true,
+				Autodiscover: []config.AutodiscoverControlPlane{
+					autodiscovery[controlplane.APIServer],
+				},
+			},
+			ControllerManager: config.ControlPlaneComponent{
+				Enabled: true,
+				Autodiscover: []config.AutodiscoverControlPlane{
+					autodiscovery[controlplane.ControllerManager],
+				},
+			},
+			Scheduler: config.ControlPlaneComponent{
+				Enabled: true,
+				Autodiscover: []config.AutodiscoverControlPlane{
+					autodiscovery[controlplane.Scheduler],
+				},
+			},
+		},
+	}
+}
+
+func createControlPlainPods(
+	t *testing.T,
+	client *fake.Clientset,
+	autodiscovery map[controlplane.ComponentName]config.AutodiscoverControlPlane,
+	nodeName string,
+) {
+	t.Helper()
+
+	for componentName, autodiscovery := range autodiscovery {
+		createControlPlainPod(t, client, componentName, autodiscovery, nodeName)
+	}
+
+	time.Sleep(time.Second)
+}
+
+func createControlPlainPod(
+	t *testing.T,
+	client *fake.Clientset,
+	componentName controlplane.ComponentName,
+	autodiscovery config.AutodiscoverControlPlane,
+	nodeName string,
+) {
+	t.Helper()
+	labelsSet, _ := labels.ConvertSelectorToLabelsMap(autodiscovery.Selector)
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      string(componentName) + fmt.Sprintf("%d", rand.Int()), // nosemgrep // add rand to allow create same component multiple times.
```

Comment:
Good ole mateusz whispered this insane trick to me: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/rand#String

Reply:
And don't forget the most important feature of kubernetes' `rand.String`: It excludes vowels and vowel-like numbers so your pods don't contain suffixes like `mydeployment-fuck4`.